nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/AbstractTrashService.java

```diff
@@ -308,33 +310,50 @@ public void untrashDocuments(List<DocumentModel> docs) {
         undeleteDocuments(docs);
     }
 
+    @Override
+    public boolean isMangledName(String docName) {
+        return TRASHED_PATTERN.matcher(docName).matches();
+    }
+
     @Override
     public String mangleName(DocumentModel doc) {
         return doc.getName() + "._" + System.currentTimeMillis() + "_.trashed";
     }
 
     @Override
     public String unmangleName(DocumentModel doc) {
-        String name = doc.getName();
-        Matcher matcher = TRASHED_PATTERN.matcher(name);
-        if (matcher.matches() && matcher.group(1).length() > 0) {
-            name = matcher.group(1);
-            matcher = COLLISION_PATTERN.matcher(name);
-            if (matcher.matches() && matcher.group(1).length() > 0) {
-                CoreSession session = doc.getCoreSession();
-                if (session != null) {
-                    String orig = matcher.group(1);
-                    String parentPath = session.getDocument(doc.getParentRef()).getPathAsString();
-                    if (parentPath.equals(PATH_SEPARATOR)) {
-                        parentPath = ""; // root
-                    }
-                    String newPath = parentPath + PATH_SEPARATOR + orig;
-                    if (!session.exists(new PathRef(newPath))) {
-                        name = orig;
-                    }
-                }
+        return unmangleName(doc.getCoreSession(), doc.getParentRef(), doc.getName());
+    }
+
+    @Override
+    public String unmangleName(CoreSession session, DocumentRef parentRef, String docName) {
+        if (session == null) {
+            return docName;
+        }
+        // remove trashed part
+        String name = getFirstGroup(TRASHED_PATTERN, docName);
+        // remove collision avoidance
+        String orig = getFirstGroup(COLLISION_PATTERN, name);
+        if (!docName.equals(name)) {
+            //check if orig (the supposed name without collision) is available
+            String parentPath = session.getDocument(parentRef).getPathAsString();
+            if (parentPath.equals(PATH_SEPARATOR)) {
+                parentPath = ""; // root
+            }
+            String newPath = parentPath + PATH_SEPARATOR + orig;
+            if (!session.exists(new PathRef(newPath))) {
+                name = orig;
             }
         }
         return name;
     }
+
+    protected String getFirstGroup(Pattern pattern, String name) {
+        Matcher matcher = pattern.matcher(name);
+        if (matcher.matches() && StringUtils.isNotEmpty(matcher.group(1))) {
+            return matcher.group(1);
+        }
+
+        return name;
+    }
```

Comment:
We refactored the unmangling method to make it easier to follow. WDYT @kevinleturc ?

Reply:
Looks good to me